<a href="https://colab.research.google.com/github/educhel/L02_Data_Analysis/blob/main/0108_%EC%8B%A4%EC%8A%B5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [1] Google Colab 알아보기

In [ ]:
users = [
    {"user_id":1,"age":25,"country":"KR","is_paid":False},
    {"user_id":2,"age":32,"country":"KR","is_paid":True},
    {"user_id":3,"age":29,"country":"US","is_paid":False},
    {"user_id":4,"age":41,"country":"US","is_paid":True},
    {"user_id":5,"age":36,"country":"KR","is_paid":True},
]

> 컨테이너 자료형 복습
- 리스트 : 값이 연속된 공간에 순서가 있게 저장되어 있는 자료형
    - 순서O, 변경O, 중복O
- 딕셔너리 : key-value의 한 쌍
    - 순서X, 변경O, 중복X(key에 대한 중복)

In [ ]:
type(users)

list

In [ ]:
users[0] # 값에 대한 접근을 위치(숫자)로 한다.
# 인덱스는 0부터 n-1 존재

{'user_id': 1, 'age': 25, 'country': 'KR', 'is_paid': False}

In [ ]:
type(users[0])

dict

In [ ]:
users[0]['age'] # value에 대한 접근을 key로 한다.

25

## (1) 조건에 따른 데이터 확인

In [ ]:
# 유료 사용자(is_paid == True)의 나이만 출력해 보세요.

# 컨테이너 반복 -> users <list> 의 요소 <dict>를 하나씩 받아서,
# 조건 비교 -> <dict>의 특정 값 (is_paid) 비교
# 비교 만족시, 나이 출력 -> if 거친 후, 출력

for user in users:
    if user['is_paid'] == True:
    # if user['is_paid']:
        print(user['age'])

32
41
36


In [ ]:
# 나이가 30세 이상인 사용자의 수를 출력해 보세요.

# 조건 : age >= 30
cnt = 0

for user in users:
    if user['age'] >= 30:
        cnt += 1

print(cnt)

3


## (2) 간단한 요약 계산

> 기술통계
- 중심 경향치 중, 하나인 "평균" 계산

In [ ]:
# 전체 사용자의 평균 나이를 계산해 보세요.

total_age = 0

for user in users:
     total_age += user['age']

avg_age = total_age / len(users)
avg_age

32.6

In [ ]:
# 유료 사용자만의 평균 나이를 계산해 보세요.

# 조건이 하나 추가!
paid_users_age = []

for user in users:
    if user["is_paid"]:
        paid_users_age.append(user['age'])

paid_avg_age = sum(paid_users_age) / len(paid_users_age)
paid_avg_age

36.333333333333336

## (3) 분석 결과 정리용 숫자 산출

In [ ]:
len(users)

5

- 전체 사용자 수는 몇 명인가요? 5명

In [ ]:
len(paid_users_age)

3

- 유료 사용자는 몇 명인가요? 3명

In [ ]:
100*(len(paid_users_age) / len(users))

60.0

- 두 집단의 나이 평균에 차이가 있나요?
    - 전체 나이 32.6 세
    - 유료사용자의 평균 나이 36.3세

Hint! 구매력이 있는 / 구매 의사가 높은 그룹 = 나이 연관이 있을 것이다.  
➡️ 가설 설정에서 활용 가능

In [ ]:
avg_age, paid_avg_age

(32.6, 36.333333333333336)

> "상황 설명"   
=> 평균만 가지고, 데이터 전체를 확인하기 어렵기 때문에 추가적인 기술 통계량 필요

# [2] Pandas : 데이터 분석 시작하기

pd.DataFrame으로 데이터 프레임 제작

- 방법 1 : 딕셔너리로 제작
- 방법 2 : 2차원 리스트로 제작

In [ ]:
# 딕셔너리로 제작
# key = 열이 됨

data = {'user_id':list(range(1, 6)),
        'age':[25, 32, 29, 41, 36],
        'country':['KR','KR','US','US','KR'],
        'is_paid':[False, True, False, True, True]}

data

{'user_id': [1, 2, 3, 4, 5],
 'age': [25, 32, 29, 41, 36],
 'country': ['KR', 'KR', 'US', 'US', 'KR'],
 'is_paid': [False, True, False, True, True]}

In [ ]:
import pandas as pd

df = pd.DataFrame(data)

In [ ]:
# 데이터프레임 형태일 때, 데이터 필터링 등이 쉽게 가능함
# 1/9 살펴볼 내용 미리보기
df.loc[df['age']>=30]

user_id  age country  is_paid
1        2   32      KR     True
3        4   41      US     True
4        5   36      KR     True

In [ ]:
df

user_id  age country  is_paid
0        1   25      KR    False
1        2   32      KR     True
2        3   29      US    False
3        4   41      US     True
4        5   36      KR     True

In [ ]:
# 방법 2: 이차원 리스트
# 모양 그대로 작성

data_list = [
    [1, 25, 'KR',False],
    [2, 32, 'KR',True],
    [3, 29, 'US',False],
    [4, 41, 'US',True],
    [5, 36, 'KR',True]
]

In [ ]:
data_list

[[1, 25, 'KR', False],
 [2, 32, 'KR', True],
 [3, 29, 'US', False],
 [4, 41, 'US', True],
 [5, 36, 'KR', True]]

In [ ]:
# 무기명의 열들이 생김
# 리스트는 이름으로 값을 관리하지 않기 때문에
pd.DataFrame(data_list)

0   1   2      3
0  1  25  KR  False
1  2  32  KR   True
2  3  29  US  False
3  4  41  US   True
4  5  36  KR   True

In [ ]:
df2 = pd.DataFrame(data_list, columns=['user_id','age','country','is_paid'])
df2

user_id  age country  is_paid
0        1   25      KR    False
1        2   32      KR     True
2        3   29      US    False
3        4   41      US     True
4        5   36      KR     True

In [ ]:
# 특이사항

# 방법 1 : 딕셔너리
data = {'user_id':list(range(1, 6)),
        'age':[25, 32, 29, 41, 36],
        'country':['KR','KR','US','US','KR'],
        'is_paid':[False, True, False, True, True, False]} # False 하나 더 추가

pd.DataFrame(data)

ValueError: All arrays must be of the same length

In [ ]:
# 방법 2: 이차원 리스트
data_list = [
    [1, 25, 'KR',False],
    [2, 32, 'KR',True],
    [3, 29, 'US',False],
    [4, 41, 'US',True],
    [5, 36, 'KR',True, True]
]
pd.DataFrame(data_list)

# => 에러가 생기지 않았다고 해서, 나의 의도대로 제작된 것은 아닐 수 있다.

0   1   2      3     4
0  1  25  KR  False  None
1  2  32  KR   True  None
2  3  29  US  False  None
3  4  41  US   True  None
4  5  36  KR   True  True

# [3] 데이터 불러오기

In [ ]:
from google.colab import drive
drive.mount('/content/drive/', force_remount = True)

Mounted at /content/drive/


In [ ]:
titatnic = pd.read_csv('/content/drive/MyDrive/새싹2기/데이터분석/data/titanic.csv')
titatnic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [ ]:
type(titatnic)

pandas.core.frame.DataFrame

In [ ]:
# 데이터 파일을 excel 형태로 내보내세요.
# 이 때, 파일명을 titanic_excel.xlsx 로, 시트명을 “data”로 지정해 주세요.
path = '/content/drive/MyDrive/새싹2기/데이터분석/data/'
file_name = 'titanic_excel.xlsx'

titatnic.to_excel(path + file_name, sheet_name='data')

In [ ]:
titatnic.to_excel(path + file_name, sheet_name='data',index=False)

# [4] 데이터 확인하기

드라이브 상 csv 파일 -> 데이터프레임 형태로 불러오기

In [ ]:
# 데이터 불러오기

# step1 : 드라이브 연결

from google.colab import drive
drive.mount('/content/drive/', force_remount=True)

Mounted at /content/drive/


In [ ]:
# step 2 : 파일 가져오기 -> 경로로 접근
# 경로 복사하기

path = '/content/drive/MyDrive/새싹2기/데이터분석/data/titanic.csv'

In [ ]:
# step 3 : 경로를 기준으로 파일 불러오기

import pandas as pd

titanic = pd.read_csv(path)

## (1) 미리보기

In [ ]:
titanic.head(10)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0         0       3    male  22.0      1      0   7.2500        S   Third   
1         1       1  female  38.0      1      0  71.2833        C   First   
2         1       3  female  26.0      0      0   7.9250        S   Third   
3         1       1  female  35.0      1      0  53.1000        S   First   
4         0       3    male  35.0      0      0   8.0500        S   Third   
5         0       3    male   NaN      0      0   8.4583        Q   Third   
6         0       1    male  54.0      0      0  51.8625        S   First   
7         0       3    male   2.0      3      1  21.0750        S   Third   
8         1       3  female  27.0      0      2  11.1333        S   Third   
9         1       2  female  14.0      1      0  30.0708        C  Second   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
5    man        True  NaN   Queenstown    no   True  
6    man        True    E  Southampton    no   True  
7  child       False  NaN  Southampton    no  False  
8  woman       False  NaN  Southampton   yes  False  
9  child       False  NaN    Cherbourg   yes  False

## (2) 구조

In [ ]:
# 전체 구조
titanic.shape

(891, 15)

In [ ]:
# 컬럼 구조
titanic.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [ ]:
len(titanic.columns)

15

## (3) 정렬하기

In [ ]:
titanic.dtypes

survived         int64
pclass           int64
sex             object
age            float64
sibsp            int64
parch            int64
fare           float64
embarked        object
class           object
who             object
adult_male        bool
deck            object
embark_town     object
alive           object
alone             bool
dtype: object

In [ ]:
# 값 정렬
# 기준 열 2개 -> class, age
# 정렬 형태 2개 -> 오름, 내림

# 단순히 정렬된 새로운 데이터 프레임 반환
# 기존 데이터 프레임을 바꾸지 않는다.
titanic.sort_values(by=['class','age'],
                    ascending=[True,False])

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
630         1       1    male  80.0      0      0  30.0000        S  First   
96          0       1    male  71.0      0      0  34.6542        C  First   
493         0       1    male  71.0      0      0  49.5042        C  First   
745         0       1    male  70.0      1      1  71.0000        S  First   
54          0       1    male  65.0      0      1  61.9792        C  First   
..        ...     ...     ...   ...    ...    ...      ...      ...    ...   
859         0       3    male   NaN      0      0   7.2292        C  Third   
863         0       3  female   NaN      8      2  69.5500        S  Third   
868         0       3    male   NaN      0      0   9.5000        S  Third   
878         0       3    male   NaN      0      0   7.8958        S  Third   
888         0       3  female   NaN      1      2  23.4500        S  Third   

       who  adult_male deck  embark_town alive  alone  
630    man        True    A  Southampton   yes   True  
96     man        True    A    Cherbourg    no   True  
493    man        True  NaN    Cherbourg    no   True  
745    man        True    B  Southampton    no  False  
54     man        True    B    Cherbourg    no  False  
..     ...         ...  ...          ...   ...    ...  
859    man        True  NaN    Cherbourg    no   True  
863  woman       False  NaN  Southampton    no  False  
868    man        True  NaN  Southampton    no   True  
878    man        True  NaN  Southampton    no   True  
888  woman       False  NaN  Southampton    no  False  

[891 rows x 15 columns]

In [ ]:
titanic.sort_values(by=['class','age'],
                    ascending=[True,False],
                    inplace=True) # 자체를 바꿔 버림

In [ ]:
# 만약! 원래대로 돌아가고 싶다면?
# -> 되돌리기는 존재하지 않습니다.
# -> 다시 특정 기준을 가지고 정렬

In [ ]:
titanic.sort_index(inplace=True)

In [ ]:
titanic.tail()

survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True